src/core/lib/event_engine/thread_pool/work_stealing_thread_pool.cc

```diff
@@ -0,0 +1,459 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/event_engine/thread_pool/work_stealing_thread_pool.h"
+
+#include <atomic>
+#include <memory>
+#include <utility>
+
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
+
+#include <grpc/support/log.h>
+
+#include "src/core/lib/backoff/backoff.h"
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/event_engine/common_closures.h"
+#include "src/core/lib/event_engine/thread_local.h"
+#include "src/core/lib/event_engine/trace.h"
+#include "src/core/lib/event_engine/work_queue/basic_work_queue.h"
+#include "src/core/lib/event_engine/work_queue/work_queue.h"
+#include "src/core/lib/gprpp/thd.h"
+#include "src/core/lib/gprpp/time.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+namespace {
+constexpr grpc_core::Duration kIdleThreadLimit =
+    grpc_core::Duration::Seconds(20);
+constexpr grpc_core::Duration kTimeBetweenThrottledThreadStarts =
+    grpc_core::Duration::Seconds(1);
+constexpr grpc_core::Duration kWorkerThreadMinSleepBetweenChecks{
+    grpc_core::Duration::Milliseconds(33)};
+constexpr grpc_core::Duration kWorkerThreadMaxSleepBetweenChecks{
+    grpc_core::Duration::Seconds(3)};
+constexpr grpc_core::Duration kLifeguardMinSleepBetweenChecks{
+    grpc_core::Duration::Milliseconds(50)};
+constexpr grpc_core::Duration kLifeguardMaxSleepBetweenChecks{
+    grpc_core::Duration::Seconds(1)};
+constexpr absl::Duration kSleepBetweenQuiesceCheck{absl::Milliseconds(10)};
+}  // namespace
+
+thread_local WorkQueue* g_local_queue = nullptr;
+
+// -------- WorkStealingThreadPool --------
+
+WorkStealingThreadPool::WorkStealingThreadPool(size_t reserve_threads)
+    : pool_{std::make_shared<WorkStealingThreadPoolImpl>(reserve_threads)} {
+  pool_->Start();
+}
+
+void WorkStealingThreadPool::Quiesce() { pool_->Quiesce(); }
+
+WorkStealingThreadPool::~WorkStealingThreadPool() {
+  GPR_ASSERT(pool_->IsQuiesced());
+}
+
+void WorkStealingThreadPool::Run(absl::AnyInvocable<void()> callback) {
+  Run(SelfDeletingClosure::Create(std::move(callback)));
+}
+
+void WorkStealingThreadPool::Run(EventEngine::Closure* closure) {
+  pool_->Run(closure);
+}
+
+// -------- WorkStealingThreadPool::TheftRegistry --------
+
+void WorkStealingThreadPool::TheftRegistry::Enroll(WorkQueue* queue) {
+  grpc_core::MutexLock lock(&mu_);
+  queues_.emplace(queue);
+}
+
+void WorkStealingThreadPool::TheftRegistry::Unenroll(WorkQueue* queue) {
+  grpc_core::MutexLock lock(&mu_);
+  queues_.erase(queue);
+}
+
+EventEngine::Closure* WorkStealingThreadPool::TheftRegistry::StealOne() {
+  grpc_core::MutexLock lock(&mu_);
+  EventEngine::Closure* closure;
+  for (auto* queue : queues_) {
```

Comment:
The bucketing idea is interesting. We could also institute queue priorities, I've seen some good results there. I just wonder if the overhead of queue juggling would negate the benefits of optimizing for queue time.

> One potential future optimization here could be that if we have a data structure that returns the queue with the highest backlog (in-terms of queue length), it might be preferable to steal from that queue first.

Maybe so. In my chats with @soheilhy, optimizing for queue time was not terribly fruitful in their experiments. Further, this is pop-most-recent because we have a high performance queue implementation that avoids mutexes for LIFO operations. It has some very rare atomic flake issues, so I planned to land that as a subsequent improvement.

But these are all things we can experiment with if the performance difference is meaningful.